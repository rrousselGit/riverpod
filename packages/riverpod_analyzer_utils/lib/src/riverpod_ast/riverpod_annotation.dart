part of '../riverpod_ast.dart';

class RiverpodAnnotationDependency extends RiverpodAst {
  RiverpodAnnotationDependency._({
    required this.node,
    required this.provider,
  });

  final Expression node;
  final GeneratorProviderDeclarationElement provider;

  @override
  void accept(RiverpodAstVisitor visitor) {
    visitor.visitRiverpodAnnotationDependency(this);
  }

  @override
  void visitChildren(RiverpodAstVisitor visitor) {}
}

class RiverpodAnnotationDependencies extends RiverpodAst {
  RiverpodAnnotationDependencies._({
    required this.node,
    required this.dependencies,
  });

  final NamedExpression node;
  final List<RiverpodAnnotationDependency>? dependencies;

  @override
  void accept(RiverpodAstVisitor visitor) {
    visitor.visitRiverpodAnnotationDependencies(this);
  }

  @override
  void visitChildren(RiverpodAstVisitor visitor) {
    final dependencies = this.dependencies;
    if (dependencies != null) {
      for (final dependency in dependencies) {
        dependency.accept(visitor);
      }
    }
  }
}

class RiverpodAnnotation extends RiverpodAst {
  RiverpodAnnotation._({
    required this.annotation,
    required this.element,
    required this.keepAliveNode,
    required this.dependencies,
  });

  static RiverpodAnnotation? _parse(
    Declaration node,
  ) {
    final annotatedElement = node.declaredElement;
    if (annotatedElement == null) return null;

    for (final annotation in node.metadata) {
      final elementAnnotation = annotation.elementAnnotation;
      final annotationElement = annotation.element;
      if (elementAnnotation == null || annotationElement == null) continue;
      if (annotationElement is! ExecutableElement ||
          !riverpodType.isExactlyType(annotationElement.returnType)) {
        // The annotation is not an @Riverpod
        continue;
      }

      final dartObject = elementAnnotation.computeConstantValue();
      if (dartObject == null) return null;

      NamedExpression? keepAliveNode;
      NamedExpression? dependenciesNode;
      final argumentList = annotation.arguments;
      if (argumentList != null) {
        for (final argument
            in argumentList.arguments.whereType<NamedExpression>()) {
          switch (argument.name.label.name) {
            case 'keepAlive':
              keepAliveNode = argument;
            case 'dependencies':
              dependenciesNode = argument;
          }
        }
      }

      final riverpodAnnotationElement =
          RiverpodAnnotationElement.parse(annotatedElement);
      if (riverpodAnnotationElement == null) return null;

      final dependencies = _parseDependencies(dependenciesNode);

      final riverpodAnnotation = RiverpodAnnotation._(
        annotation: annotation,
        element: riverpodAnnotationElement,
        keepAliveNode: keepAliveNode,
        dependencies: dependencies,
      );
      dependencies?._parent = riverpodAnnotation;

      return riverpodAnnotation;
    }

    return null;
  }

  static RiverpodAnnotationDependencies? _parseDependencies(
    NamedExpression? dependenciesNode,
  ) {
    if (dependenciesNode == null) return null;
    final dependenciesNodeValue = dependenciesNode.expression;
    // TODO handle Riverpod(dependencies:null)

    final dependencies = <RiverpodAnnotationDependency>[];

    if (dependenciesNodeValue is! ListLiteral) {
      errorReporter?.call(
        RiverpodAnalysisError(
          '@Riverpod(dependencies: <...>) only support list literals (using []).',
          targetNode: dependenciesNodeValue,
        ),
      );
    } else {
      for (final dependency in dependenciesNodeValue.elements) {
        if (dependency is! Expression) {
          errorReporter?.call(
            RiverpodAnalysisError(
              '@Riverpod(dependencies: [...]) does not support if/for/spread operators.',
              targetNode: dependency,
            ),
          );
          continue;
        }

        if (dependency is! SimpleIdentifier) {
          errorReporter?.call(
            RiverpodAnalysisError(
              'Only elements annotated with @riverpod are supported as "dependencies".',
              targetNode: dependency,
            ),
          );
          continue;
        }

        final dependencyElement = dependency.staticElement;
        if (dependencyElement is FunctionElement) {
          final dependencyProvider = FunctionalProviderDeclarationElement.parse(
            dependencyElement,
            annotation: null,
          );
          if (dependencyProvider == null) {
            errorReporter?.call(
              RiverpodAnalysisError(
                'The dependency $dependency is not a class annotated with @riverpod',
                targetNode: dependency,
              ),
            );
            continue;
          }

          dependencies.add(
            RiverpodAnnotationDependency._(
              node: dependency,
              provider: dependencyProvider,
            ),
          );
        } else if (dependencyElement is ClassElement) {
          final dependencyProvider = ClassBasedProviderDeclarationElement.parse(
            dependencyElement,
            annotation: null,
          );
          if (dependencyProvider == null) {
            errorReporter?.call(
              RiverpodAnalysisError(
                'The dependency $dependency is not a class annotated with @riverpod',
                targetNode: dependency,
              ),
            );
            continue;
          }

          dependencies.add(
            RiverpodAnnotationDependency._(
              node: dependency,
              provider: dependencyProvider,
            ),
          );
        } else {
          errorReporter?.call(
            RiverpodAnalysisError(
              '@Riverpod(dependencies: [...]) only supports elements annotated with @riverpod as values.',
              targetNode: dependency,
            ),
          );
        }
      }
    }

    final riverpodAnnotationDependencies = RiverpodAnnotationDependencies._(
      node: dependenciesNode,
      dependencies: dependencies,
    );

    for (final dependency in dependencies) {
      dependency._parent = riverpodAnnotationDependencies;
    }

    return riverpodAnnotationDependencies;
  }

  final Annotation annotation;
  final RiverpodAnnotationElement element;
  final NamedExpression? keepAliveNode;
  final RiverpodAnnotationDependencies? dependencies;

  @override
  void accept(RiverpodAstVisitor visitor) {
    visitor.visitRiverpodAnnotation(this);
  }

  @override
  void visitChildren(RiverpodAstVisitor visitor) {
    dependencies?.accept(visitor);
  }
}
