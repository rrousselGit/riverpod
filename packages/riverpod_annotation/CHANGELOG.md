## 2.3.4 - 2024-02-03

- Improved `@Riverpod(dependencies: [...])` documentation.

## 2.3.3 - 2023-11-27

- `riverpod` upgraded to `2.4.9`

## 2.3.2 - 2023-11-20

- `riverpod` upgraded to `2.4.8`

## 2.3.1 - 2023-11-20

- `riverpod` upgraded to `2.4.7`

## 2.3.1 - 2023-11-13

- `riverpod` upgraded to `2.4.6`

## 2.3.0 - 2023-10-28

- Exported internal `FamilyOverride` API, for use in generated code.

## 2.2.1 - 2023-10-15

- `riverpod` upgraded to `2.4.4`

## 2.2.0 - 2023-10-06

- Exports `@internal` from `pkg:meta` for the code-generator to use.

## 2.1.6 - 2023-09-27

- `riverpod` upgraded to `2.4.1`

## 2.1.5 - 2023-09-04

- `riverpod` upgraded to `2.4.0`

## 2.1.4 - 2023-08-28

- `riverpod` upgraded to `2.3.10`

## 2.1.3 - 2023-08-28

- `riverpod` upgraded to `2.3.8`

## 2.1.2 - 2023-08-16

- `riverpod` upgraded to `2.3.7`

## 2.1.1 - 2023-04-24

- `riverpod` upgraded to `2.3.6`

## 2.1.0 - 2023-04-18

- Added support for `Raw` typedef in the return value of providers.
  This can be used to disable the conversion of Futures/Streams into AsyncValues
  ```dart
  @riverpod
  Raw<Future<int>> myProvider(...) async => ...;
  ...
  // returns a Future<int> instead of AsyncValue<int>
  Future<int> value = ref.watch(myProvider);
  ```

## 2.0.4 - 2023-04-07

- `riverpod` upgraded to `2.3.4`

## 2.0.3 - 2023-04-06

- `riverpod` upgraded to `2.3.3`

## 2.0.2 - 2023-03-13

- `riverpod` upgraded to `2.3.2`

## 2.0.1 - 2023-03-09

- `riverpod` upgraded to `2.3.1`

## 2.0.0

- Export necessary utilities for providers returning a Stream.
- Upgraded riverpod dependency

## 1.2.1

Bump minimum Riverpod version

## 1.2.0

- It is now possible to specify `@Riverpod(dependencies: [...])` to scope providers
- Marked `@Riverpod` as `@sealed`

## 1.1.1

Upgrade Riverpod to latest

## 1.1.0

Upgrade Riverpod to latest

## 1.0.6

Upgrade Riverpod to latest

## 1.0.5

Upgrade Riverpod to latest

## 1.0.4

Export more missing types

## 1.0.3

Export missing types

## 1.0.2

- Update a dependency to the latest release.

## 1.0.1

Upgrade Riverpod version

## 1.0.0

Initial release
