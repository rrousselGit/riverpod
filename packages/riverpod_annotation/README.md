<p align="center">
<a href="https://github.com/rrousselGit/riverpod/actions"><img src="https://github.com/rrousselGit/riverpod/workflows/Build/badge.svg" alt="Build Status"></a>
<a href="https://codecov.io/gh/rrousselgit/riverpod"><img src="https://codecov.io/gh/rrousselgit/riverpod/branch/master/graph/badge.svg" alt="codecov"></a>
<a href="https://github.com/rrousselgit/riverpod"><img src="https://img.shields.io/github/stars/rrousselgit/riverpod.svg?style=flat&logo=github&colorB=deeppink&label=stars" alt="Star on Github"></a>
<a href="https://opensource.org/licenses/MIT"><img src="https://img.shields.io/badge/license-MIT-purple.svg" alt="License: MIT"></a>
<a href="https://discord.gg/Bbumvej"><img src="https://img.shields.io/discord/765557403865186374.svg?logo=discord&color=blue" alt="Discord"></a>

<p align="center">
<a href="https://www.netlify.com">
  <img src="https://www.netlify.com/img/global/badges/netlify-color-accent.svg" alt="Deploys by Netlify" />
</a>
</p>

<p align="center">
<img src="https://github.com/rrousselGit/riverpod/blob/master/resources/icon/Facebook%20Cover%20A.png?raw=true" width="100%" alt="Riverpod" />
</p>

</p>

---

# Riverpod annotation

This is a side package for riverpod_generator, exposing annotations

See [riverpod_generator](https://pub.dev/packages/riverpod_generator) for more information.
