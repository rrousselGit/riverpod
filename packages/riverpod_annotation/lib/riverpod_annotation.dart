export 'dart:async' show FutureOr;

// Annotations used by code-generators
export 'package:meta/meta.dart' show visibleForOverriding;

// ignore: invalid_export_of_internal_element
export 'package:riverpod/src/internals.dart'
    show
        // General stuff
        Family,
        ProviderOrFamily,
        Override,
        // ignore: invalid_use_of_internal_member, Used by notifiers for overriding overrideWith
        ProviderOverride,
        // ignore: invalid_use_of_internal_member, used by families for overrideWith
        FamilyOverride,

        // Provider
        Provider,
        ProviderFamily,
        ProviderRef,
        AutoDisposeProvider,
        AutoDisposeProviderFamily,
        AutoDisposeProviderRef,
        ProviderElement,
        AutoDisposeProviderElement,

        // FutureProvider
        FutureProvider,
        FutureProviderFamily,
        FutureProviderRef,
        AutoDisposeFutureProvider,
        AutoDisposeFutureProviderFamily,
        AutoDisposeFutureProviderRef,
        FutureProviderElement,
        AutoDisposeFutureProviderElement,

        // StreamProvider
        StreamProvider,
        StreamProviderFamily,
        StreamProviderRef,
        AutoDisposeStreamProvider,
        AutoDisposeStreamProviderFamily,
        AutoDisposeStreamProviderRef,
        StreamProviderElement,
        AutoDisposeStreamProviderElement,

        // AsyncValue
        AsyncValue,
        AsyncLoading,
        AsyncData,
        AsyncError,
        AsyncValueX,

        // Notifier
        Notifier,
        AutoDisposeNotifier,
        NotifierProviderElement,
        AutoDisposeNotifierProviderElement,
        // ignore: invalid_use_of_internal_member
        NotifierProviderImpl,
        // ignore: invalid_use_of_internal_member
        AutoDisposeNotifierProviderImpl,
        // ignore: invalid_use_of_internal_member
        BuildlessNotifier,
        // ignore: invalid_use_of_internal_member
        BuildlessAutoDisposeNotifier,
        NotifierProvider,
        NotifierProviderRef,
        AutoDisposeNotifierProvider,
        AutoDisposeNotifierProviderRef,

        // AsyncNotifier
        AsyncNotifier,
        AutoDisposeAsyncNotifier,
        AsyncNotifierProviderElement,
        AutoDisposeAsyncNotifierProviderElement,
        // ignore: invalid_use_of_internal_member
        AsyncNotifierProviderImpl,
        // ignore: invalid_use_of_internal_member
        AutoDisposeAsyncNotifierProviderImpl,
        // ignore: invalid_use_of_internal_member
        BuildlessAsyncNotifier,
        // ignore: invalid_use_of_internal_member
        BuildlessAutoDisposeAsyncNotifier,
        AsyncNotifierProvider,
        AsyncNotifierProviderRef,
        AutoDisposeAsyncNotifierProvider,
        AutoDisposeAsyncNotifierProviderRef,

        // StreamNotifier
        StreamNotifier,
        AutoDisposeStreamNotifier,
        StreamNotifierProviderElement,
        AutoDisposeStreamNotifierProviderElement,
        // ignore: invalid_use_of_internal_member
        StreamNotifierProviderImpl,
        // ignore: invalid_use_of_internal_member
        AutoDisposeStreamNotifierProviderImpl,
        // ignore: invalid_use_of_internal_member
        BuildlessStreamNotifier,
        // ignore: invalid_use_of_internal_member
        BuildlessAutoDisposeStreamNotifier,
        StreamNotifierProvider,
        StreamNotifierProviderRef,
        AutoDisposeStreamNotifierProvider,
        AutoDisposeStreamNotifierProviderRef;

export 'src/riverpod_annotation.dart';
