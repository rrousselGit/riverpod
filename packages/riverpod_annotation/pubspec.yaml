name: riverpod_annotation
description: A package exposing annotations for riverpod_generator
version: 2.3.4
repository: https://github.com/rrousselGit/riverpod
issue_tracker: https://github.com/rrousselGit/riverpod/issues
funding:
  - https://github.com/sponsors/rrousselGit/

environment:
  sdk: ">=2.17.0 <3.0.0"

dependencies:
  meta: ^1.7.0
  riverpod: ^2.5.0

dev_dependencies:
  test: ^1.21.0
