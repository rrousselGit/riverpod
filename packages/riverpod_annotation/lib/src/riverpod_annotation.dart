// DO NOT MOVE/RENAME

import 'package:meta/meta.dart';
import 'package:meta/meta_meta.dart';

import '../riverpod_annotation.dart';

/// {@template riverpod_annotation.provider}
/// An annotation placed on classes or functions.
///
/// This tells riverpod_generator to generate a provider out of the annotated.
///
/// By default, Riverpod will convert [Future]s and [Stream]s into [AsyncValue]s.
/// If this is undesired, you can use [Raw] to have Riverpod forcibly return
/// the raw [Future]/[Stream] instead.
/// element.
/// {@endtemplate}
@Target({TargetKind.classType, TargetKind.function})
@sealed
class Riverpod {
  /// {@macro riverpod_annotation.provider}
  const Riverpod({
    this.keepAlive = false,
    this.dependencies,
  });

  /// Whether the state of the provider should be maintained if it is no-longer used.
  ///
  /// Defaults to false.
  final bool keepAlive;

  /// The list of providers that this provider potentially depends on.
  ///
  /// This list must contains the classes/functions annotated with `@riverpod`,
  /// not the generated providers themselves.
  ///
  /// Specifying this list is strictly equivalent to saying "This provider may
  /// be scoped". If a provider is scoped, it should specify [dependencies].
  /// If it is never scoped, it should not specify [dependencies].
  ///
  /// The content of [dependencies] should be a list of all the providers that
  /// this provider may depend on which can be scoped.
  ///
  /// For example, consider the following providers:
  /// ```dart
  /// // By not specifying "dependencies", we are saying that this provider is never scoped
  /// @riverpod
  /// Foo root(RootRef ref) => Foo();
  /// // By specifying "dependencies" (even if the list is empty),
  /// // we are saying that this provider is potentially scoped
  /// @Riverpod(dependencies: [])
  /// Foo scoped(ScopedRef ref) => Foo();
  /// ```
  ///
  /// Then if we were to depend on `rootProvider` in a scoped provider, we
  /// could write any of:
  ///
  /// ```dart
  /// @riverpod
  /// Object? dependent(DependentRef ref) {
  ///   ref.watch(rootProvider);
  ///   // This provider does not depend on any scoped provider,
  ///   // as such "dependencies" is optional
  /// }
  ///
  /// @Riverpod(dependencies: [])
  /// Object? dependent(DependentRef ref) {
  ///   ref.watch(rootProvider);
  ///   // This provider decided to specify "dependencies" anyway, marking
  ///   // "dependentProvider" as possibly scoped.
  ///   // Since "rootProvider" is never scoped, it doesn't need to be included
  ///   // in "dependencies".
  /// }
  ///
  /// @Riverpod(dependencies: [root])
  /// Object? dependent(DependentRef ref) {
  ///   ref.watch(rootProvider);
  ///   // Including "rootProvider" in "dependencies" is fine too, even though
  ///   // it is not required. It is a no-op.
  /// }
  /// ```
  ///
  /// However, if we were to depend on `scopedProvider` then our only choice is:
  ///
  /// ```dart
  /// @Riverpod(dependencies: [scoped])
  /// Object? dependent(DependentRef ref) {
  ///   ref.watch(scopedProvider);
  ///   // Since "scopedProvider" specifies "dependencies", any provider that
  ///   // depends on it must also specify "dependencies" and include "scopedProvider".
  /// }
  /// ```
  ///
  /// In that scenario, the `dependencies` parameter is required and it must
  /// include `scopedProvider`.
  ///
  /// See also:
  /// - [provider_dependencies](https://github.com/rrousselGit/riverpod/tree/master/packages/riverpod_lint#provider_dependencies-riverpod_generator-only)
  ///   and [scoped_providers_should_specify_dependencies](https://github.com/rrousselGit/riverpod/tree/master/packages/riverpod_lint#scoped_providers_should_specify_dependencies-generator-only).\
  ///   These are lint rules that will warn about incorrect `dependencies` usages.
  final List<Object>? dependencies;
}

/// {@macro riverpod_annotation.provider}
@Target({TargetKind.classType, TargetKind.function})
const riverpod = Riverpod();

/// An annotation used to help the linter find the user-defined element from
/// the generated provider.
///
/// DO NOT USE
class ProviderFor {
  /// An annotation used to help the linter find the user-defined element from
  /// the generated provider.
  // Put the annotation on the constructor to avoid the linter from complaining
  // about the annotation being exported; while preventing the user from using it.
  @internal
  const ProviderFor(this.value)
      : assert(
          value is Function || value is Type,
          '$value is not a class/function',
        );

  /// The code annotated by `@riverpod`
  final Object value;
}

/// {@template riverpod_annotation.raw}
/// An annotation for marking a value type as "should not be handled
/// by Riverpod".
///
/// This is a type-alias to [T], and has no runtime effect. It is only used
/// as metadata for the code-generator/linter.
///
/// This serves two purposes:
/// - It enables a provider to return a [Future]/[Stream] without
///   having the provider converting it into an [AsyncValue].
///   ```dart
///   @riverpod
///   Raw<Future<int>> myProvider(...) async => ...;
///   ...
///   // returns a Future<int> instead of AsyncValue<int>
///   Future<int> value = ref.watch(myProvider);
///   ```
///
/// - It can silence the linter when a provider returns a value that
///   is otherwise not supported, such as Flutter's `ChangeNotifier`:
///   ```dart
///   // Will not trigger the "unsupported return type" lint
///   @riverpod
///   Raw<MyChangeNotifier> myProvider(...) => MyChangeNotifier();
///   ```
///
/// The typedef can be used at various places within the return type declaration.
///
/// For example, a valid return type is `Future<Raw<ChangeNotifier>>`.
/// This way, Riverpod will convert the [Future] into an [AsyncValue], and
/// the usage of `ChangeNotifier` will not trigger the linter:
///
/// ```dart
/// @riverpod
/// Future<Raw<ChangeNotifier>> myProvider(...) async => ...;
/// ...
/// AsyncValue<ChangeNotifier> value = ref.watch(myProvider);
/// ```
///
/// {@endtemplate}
typedef Raw<T> = T;
