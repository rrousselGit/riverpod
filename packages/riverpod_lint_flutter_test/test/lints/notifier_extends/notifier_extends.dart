import 'package:riverpod_annotation/riverpod_annotation.dart';

part 'notifier_extends.g.dart';

@riverpod
class MyNotifier extends _$MyNotifier {
  int build() => 0;
}

@riverpod
// expect_lint: notifier_extends
class NoExtends {
  int build() => 0;
}

@riverpod
// expect_lint: notifier_extends
class WrongExtends extends AsyncNotifier<int> {
  int build() => 0;
}

// Regression test for https://github.com/rrousselGit/riverpod/issues/2165
@riverpod
class _PrivateClass extends _$PrivateClass {
  @override
  String build() => 'Hello World!';
}
