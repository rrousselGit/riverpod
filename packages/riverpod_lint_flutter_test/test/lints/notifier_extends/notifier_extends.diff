Message: `Extend _$NoExtends`
Priority: 90
Diff for file `test/lints/notifier_extends/notifier_extends.dart:12`:
```
@riverpod
// expect_lint: notifier_extends
- class NoExtends {
+ class NoExtends extends _$NoExtends {
  int build() => 0;
}
```
---
Message: `Extend _$WrongExtends`
Priority: 90
Diff for file `test/lints/notifier_extends/notifier_extends.dart:18`:
```
@riverpod
// expect_lint: notifier_extends
- class WrongExtends extends AsyncNotifier<int> {
+ class WrongExtends extends _$WrongExtends {
  int build() => 0;
}
```
---
