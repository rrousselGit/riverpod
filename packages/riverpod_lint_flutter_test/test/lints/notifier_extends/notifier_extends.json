[
  {
    "priority": 90,
    "change": {
      "message": "Extend _$NoExtends",
      "edits": [
        {
          "fileStamp": 0,
          "edits": [
            {
              "offset": 228,
              "length": 0,
              "replacement": " extends _$NoExtends"
            }
          ]
        }
      ],
      "linkedEditGroups": []
    }
  },
  {
    "priority": 90,
    "change": {
      "message": "Extend _$WrongExtends",
      "edits": [
        {
          "fileStamp": 0,
          "edits": [
            {
              "offset": 324,
              "length": 18,
              "replacement": "_$WrongExtends"
            }
          ]
        }
      ],
      "linkedEditGroups": []
    }
  }
]