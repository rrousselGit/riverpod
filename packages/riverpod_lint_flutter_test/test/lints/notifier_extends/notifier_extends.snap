Message: `Extend _$NoExtends`
<<<< start: 229 -- end: 229
==== start: 229 -- end: 249
extends _$NoExtends 
>>>>
Message: `Extend _$WrongExtends`
<<<< start: 324 -- end: 341
AsyncNotifier<int
==== start: 324 -- end: 337
_$WrongExtend
>>>>
